File: cpp/homemodel.cpp
#include <qdebug.h>
#include <QQmlApplicationEngine>
#include "homemodel.h"
#include "librarymodel.h"
#include "rustutil.h"


HomeModel::HomeModel(QObject *parent, QQmlApplicationEngine *engine) : QAbstractListModel {parent} {
    HomeModel::updateLibraryList();
}
void HomeModel::addLibrary(const QString& path){
	QString name = path.split("/").last();
	beginInsertRows(QModelIndex(), libraryList.size(), libraryList.size());
	this->libraryList.push_back(add_library(name.toStdString(), path.toStdString()));
	endInsertRows();

}
void HomeModel::deleteLibrary(int row) {
	beginRemoveRows(QModelIndex(), row, row);
	delete_library(libraryList.at(row).uuid);
	updateLibraryList();
	endRemoveRows();
}

void HomeModel::updateLibraryList() {
	beginInsertRows(QModelIndex(), 0, 0);
	rust::Vec<Library> lib_vec = get_library_list();
	this->libraryList = lib_vec;
	endInsertRows();
}

int HomeModel::rowCount(const QModelIndex &parent) const {
	if (parent.isValid()) return 0;
	return this->libraryList.size();
}


QVariant HomeModel::data(const QModelIndex &index, int role) const {
	switch(role) {
		case UuidRole: return QString::fromStdString(std::string((libraryList.at(index.row()).uuid)));
		case NameRole: return QString::fromStdString(std::string((libraryList.at(index.row()).name)));
		case PathRole: return QString::fromStdString(std::string((libraryList.at(index.row()).path)));
	}
	return "";
}

QHash<int, QByteArray> HomeModel::roleNames() const {
    static QHash<int, QByteArray> mapping {
		{UuidRole, "uuid"},
		{NameRole, "name"},
		{PathRole, "path"},
	};
	return mapping;
}

void HomeModel::openLibrary(int row){
	Library library = libraryList.at(row);

	QString libraryContext = "library" + RustUtil::asQString(library.uuid);
}

File: rust/client_lib/src/home_model.rs


use crate::db::*;
use lazy_static::lazy_static;
use std::sync::Mutex;

use crate::library_model::create_library;

pub struct Library {
	pub uuid: String,
	pub name: String,
	pub path: String,
}

lazy_static!{
    static ref HOME_DB_CONN: Mutex<home_db::HomeDb> = Mutex::new(home_db::HomeDb::new());

}

pub fn add_library(name: &str, path: &str) -> Library{
	let home_db = HOME_DB_CONN.lock().unwrap();
	let local_path = path.replace("file://", "");
	let library = home_db.add_library(name, local_path.as_str());
	create_library(library.uuid.as_str(), library.path.as_str());


	return library
}

pub fn get_library_list() -> Vec<Library>{
	let home_db = HOME_DB_CONN.lock().unwrap();
	let library_list = home_db.get_libraries();
	library_list
}

pub fn delete_library(library_uuid: &str){
	let home_db = HOME_DB_CONN.lock().unwrap();
	home_db.delete_library(library_uuid);
}
